packages/nx/src/utils/cache-directory.ts

```diff
@@ -1,40 +1,106 @@
-import { join, isAbsolute } from 'path';
-import { workspaceRoot } from './workspace-root';
-import { readJsonFile } from './fileutils';
+import { isAbsolute, join } from 'path';
 import { NxJsonConfiguration } from '../config/nx-json';
+import { readJsonFile } from './fileutils';
+import { workspaceRoot } from './workspace-root';
 
-function readCacheDirectoryProperty(root: string): string | undefined {
-  try {
-    const nxJson = readJsonFile<NxJsonConfiguration>(join(root, 'nx.json'));
-    return nxJson.tasksRunnerOptions.default.options.cacheDirectory;
-  } catch {
-    return undefined;
-  }
+export interface CacheDirectorySpec {
+  envName: string;
+  propertyName: string;
+  defaultDirectory?: string;
 }
 
-function cacheDirectory(root: string, cacheDirectory: string) {
-  const cacheDirFromEnv = process.env.NX_CACHE_DIRECTORY;
-  if (cacheDirFromEnv) {
-    cacheDirectory = cacheDirFromEnv;
+/**
+ * Calculate path to cache directory
+ * The path is calculated by looking at these values in order of precedense:
+ * env[cache.envName]
+ * nxConfig.tasksRunnerOptions.default.options[cache.propertyName]
+ * cache.defaultDirectory
+ *
+ * if the resulting path from the above rule is not an absolute path
+ * it will be prefixed with root
+ *
+ * @param root : the root of the workspace
+ * @param env : the environment - e.g. process.env
+ * @param nxConfig : the content of the workspace nx.json
+ * @param cache : the specification of the cache :
+ * -              envName: the property in env that may define the cache
+ * -              propertyName: the property in
+ * -                            nxConfig.tasksRunnerOptions.default.options
+ * -                            that may define the cache
+ * -              defaultDirectory: the cache directory path if none of the
+ * -                                above defines the path
+ * @returns
+ */
+export function cacheDirectory(
+  root: string,
+  env: { [key: string]: string },
+  nxConfig: NxJsonConfiguration,
+  cache: CacheDirectorySpec
+): string {
+  let directory = env[cache.envName];
+  if (!directory) {
+    directory =
+      nxConfig.tasksRunnerOptions?.default?.options[cache.propertyName];
+  }
+  if (!directory) {
```

Comment:
I do think its easier to read after you reversed the order of the `if-else` but I would prefer to keep it as an `if-else` rather than splitting it into 2 `if` blocks. Can you change it please?

Reply:
Sorry for the delay - was somewhat busy at work
I totally agree that there is no likely need to move the project graph cache directory away from the workspace - i.e. `node_modules/.cache/nx` so I should probably create a new PR with a different title ?

As far as using `if-else` the code would look like this (I added support for unspecified `envName` and `propertyName` : 

```
export interface CacheDirectorySpec {
  envName?: string;
  propertyName?: string;
  defaultDirectory?: string;
}

/**
 * Calculate path to cache directory
 * The path is calculated by looking at these values in order of precedense:
 * env[cache.envName]
 * nxConfig.tasksRunnerOptions.default.options[cache.propertyName]
 * cache.defaultDirectory
 *
 * if the resulting path from the above rule is not an absolute path
 * it will be prefixed with root
 *
 * @param root : the root of the workspace
 * @param env : the environment - e.g. process.env
 * @param nxConfig : the content of the workspace nx.json
 * @param cache : the specification of the cache :
 * -              envName: the optional property in env that may define the cache
 * -              propertyName: the optional property in
 * -                            nxConfig.tasksRunnerOptions.default.options
 * -                            that may define the cache
 * -              defaultDirectory: the cache directory path if none of the
 * -                                above defines the path
 * @returns
 */
export function cacheDirectory(
  root: string,
  env: { [key: string]: string },
  nxConfig: NxJsonConfiguration,
  cache: CacheDirectorySpec
): string {
  let directory = '';
  if (cache.envName && env[cache.envName]) {
    directory = env[cache.envName];
  } else if (
    cache.propertyName &&
    nxConfig.tasksRunnerOptions?.default?.options[cache.propertyName]
  ) {
    directory =
      nxConfig.tasksRunnerOptions?.default?.options[cache.propertyName];
  } else {
    directory = cache.defaultDirectory;
  }

  if (directory) {
    if (isAbsolute(directory)) {
      return directory;
    } else {
      return join(root, directory);
    }
  }
  return '';
}
```